crates/bevy_ecs/macros/src/component.rs

```diff
@@ -202,11 +283,14 @@ pub fn document_required_components(attr: TokenStream, item: TokenStream) -> Tok
 pub const COMPONENT: &str = "component";
 pub const STORAGE: &str = "storage";
 pub const REQUIRE: &str = "require";
+pub const RELATIONSHIP: &str = "relationship";
+pub const RELATIONSHIP_SOURCES: &str = "relationship_sources";
 
 pub const ON_ADD: &str = "on_add";
 pub const ON_INSERT: &str = "on_insert";
 pub const ON_REPLACE: &str = "on_replace";
 pub const ON_REMOVE: &str = "on_remove";
+pub const ON_DESPAWN: &str = "on_despawn";
```

Comment:
Probably not something that should be done in this PR, but do we also want an `on_spawn` hook for completeness? There's already nice symmetry with `on_add`/`on_remove` and `on_insert`/`on_replace` (or `on_discard` if we want to rename it to that)

Reply:
Worth discussing, but I can't personally think of a spawn use case that wouldn't be better off in OnAdd or OnInsert. Given that every hook has a cost associated with it, we need real use cases to justify their existence.

Part of me thinks we might even be better off hard-coding the "recursive despawn" logic rather than using hooks, as I also can't think of a use case for on_despawn other than recursive despawn.